internal/driver/gomobile/android.c

```diff
@@ -155,93 +158,183 @@ void closeStream(uintptr_t jni_env, uintptr_t ctx, void* stream) {
-	if (getDoc == NULL) { // API 21
-		return false;
-	}
-	jstring docID = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getDoc, uri);
+	return memcmp(prefix, string, lp) == 0;
+}
 
-	jmethodID getTree = find_static_method(env, contractClass, "buildDocumentUriUsingTree", "(Landroid/net/Uri;Ljava/lang/String;)Landroid/net/Uri;");
-	jobject treeUri = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getTree, uri, docID);
+bool uriCanList(uintptr_t jni_env, uintptr_t ctx, char* uriCstr) {
+	if (hasPrefix(uriCstr, "file://")) {
+		// Get file path from URI
+		size_t length = strlen(uriCstr)-7;// -7 for 'file://'
+		char* path = malloc(sizeof(char)*(length+1));// +1 for '\0'
+		memcpy(path, &uriCstr[7], length);
+		path[length] = '\0';
 
-	jclass resolverClass = (*env)->GetObjectClass(env, resolver);
-	jmethodID getType = find_method(env, resolverClass, "getType", "(Landroid/net/Uri;)Ljava/lang/String;");
-	jstring type = (jstring)(*env)->CallObjectMethod(env, resolver, getType, treeUri);
+		// Stat path to determine if it points to a directory
+		struct stat statbuf;
+		int result = stat(path, &statbuf);
 
-	if (type == NULL) {
-		return false;
-	}
+		free(path);
+
+		if (result == 0) {
+			return S_ISDIR(statbuf.st_mode);
+		}
+	} else if (hasPrefix(uriCstr, "content://")) {
+		JNIEnv *env = (JNIEnv*)jni_env;
+		jobject resolver = getContentResolver(jni_env, ctx);
+		jstring uriStr = (*env)->NewStringUTF(env, uriCstr);
+		jobject uri = parseURI(jni_env, ctx, uriCstr);
+		jthrowable loadErr = (*env)->ExceptionOccurred(env);
+
+		if (loadErr != NULL) {
+			(*env)->ExceptionDescribe(loadErr);
+			(*env)->ExceptionClear(env);
+			return false;
+		}
+
+		jclass contractClass = find_class(env, "android/provider/DocumentsContract");
+		if (contractClass == NULL) { // API 19
+			return false;
+		}
+		jmethodID getDoc = find_static_method(env, contractClass, "getTreeDocumentId", "(Landroid/net/Uri;)Ljava/lang/String;");
+		if (getDoc == NULL) { // API 21
+			return false;
+		}
+		jstring docID = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getDoc, uri);
+
+		jmethodID getTree = find_static_method(env, contractClass, "buildDocumentUriUsingTree", "(Landroid/net/Uri;Ljava/lang/String;)Landroid/net/Uri;");
+		jobject treeUri = (jobject)(*env)->CallStaticObjectMethod(env, contractClass, getTree, uri, docID);
+
+		jclass resolverClass = (*env)->GetObjectClass(env, resolver);
+		jmethodID getType = find_method(env, resolverClass, "getType", "(Landroid/net/Uri;)Ljava/lang/String;");
+		jstring type = (jstring)(*env)->CallObjectMethod(env, resolver, getType, treeUri);
 
-	char *str = getString(jni_env, ctx, type);
-	return strcmp(str, "vnd.android.document/directory") == 0;
+		if (type == NULL) {
+			return false;
+		}
+
+		char *str = getString(jni_env, ctx, type);
+		return strcmp(str, "vnd.android.document/directory") == 0;
+	}
```

Comment:
I think we should log here that the URI scheme was not known - there will be no other errors logged to help figure such a corner case out.

Reply:
Yup good idea, should this be log_fatal, or just log_info?